### ISSUE TITLE
template/copy module regression: allow redirection inside validate

### ISSUE THREAD

**Issue by juju4** [ansible/ansible#22773]
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
template module.
can apply to copy module

##### ANSIBLE VERSION
```
$ ansible --version
ansible 2.2.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
```
+2.2.1

##### OS / ENVIRONMENT
Ubuntu trusty host or Travis with docker host (trusty/xenial/...)

##### SUMMARY
Under 2.2.0, validate arguments with shell redirection worked.
Under 2.2.1 (Travis), it does not.

##### STEPS TO REPRODUCE

```
- block:
    - name: ensure iptables directory exists
      file: dest=/etc/iptables state=directory mode=0755
    - name: iptables configuration file update
      template:
        src: "{{ harden_fw_template }}"
        dest: /etc/iptables/rules.v4
        backup: yes 
        owner: '0' 
        group: '0' 
        mode: '0644'
## works in ansible 2.2.0.0, nok 2.2.1.0?
#        validate: 'iptables-restore -tv < %s'
      notify:
        - restart iptables
        - restart iptables - xenial
      register: rulesv4
## as validate option seems incorrect
    - name: ensure rules are correct
      shell: "iptables-restore -tv < /etc/iptables/rules.v4"
      when: rulesv4.changed
  when: harden_fw is defined and harden_fw
```

##### ACTUAL RESULTS
with 2.2.1 (2.2.0 fine)
```
FAILED! => {"changed": true, "exit_status": 1, "failed": true, "msg": "failed to validate", "stderr": "Unknown arguments found on commandline\n", "stdout": "", "stdout_lines": []}

```

**Comment by ansibot** [comment]
Files identified in the description:
* [lib/ansible/modules/files/copy.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/files/copy.py)
* [lib/ansible/modules/files/template.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/files/template.py)

If these files are inaccurate, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by abadger** [comment]
Sorry for taking so long to process this bug report.

The validate option does not take shell metachars for safety reasons. I checked the documentation for the validate parameter as far back as 1.9 and it reads:

>- validate
>        The validation command to run before copying into place.  The
>        path to the file to validate is passed in via '%s' which must
>        be present as in the visudo example below. The command is
>        passed securely so shell features like expansion and pipes
>        won't work. [Default: ]

I also ran with checkouts of all releases from 1.7 onwards and it similarly failed so I was unable to reproduce 2.2.0 allowing this.  However, I do note that things could have appeared to work if iptables-restore processed the shell metachars itself and either discarded them as garbage or exited with a 0 exit status in this case.  Ansible would have run the equivalent of this shell command:

```
iptables-restore '-tv' '<' 'iptables'
```

In the version of iptables-restore that I currently have, iptables-restore complains about '<' being an unknown argument and exits with exit status 1.  But if a previous version of iptables-restore processed the ```<``` as a filename and then did not fail if the file did not exist, it would have appeared that ansible-2.2.0 successfully validated the file prior to copying it into place.

I'm going to close this as ansible is working as expected.
